### PR TITLE
(FACT-1472) prepare external facts caching

### PR DESCRIPTION
**Pull request by ciprianbadescu**

### CODE REVIEW

**Comment by ciprianbadescu** [comment]
jenkins please test this on debian9-64a,debian10-64a,fedora29-64a,fedora30-64a,osx1014-64a,redhat5-64a,centos7-64a,centos8-64a,redhat7-64a,redhat7-POWERa,sles12-64a,sles15-64a,solaris114-64a,ubuntu1604-64a,ubuntu1604-32a,ubuntu1804-64a,windows10ent-64a,windows2012r2-64a,windows2016-64a,windows2019-64a

**Comment by ciprianbadescu** [comment]
jenkins please test this on debian9-64a,debian10-64a,fedora29-64a,fedora30-64a,osx1014-64a,redhat5-64a,centos7-64a,centos8-64a,redhat7-64a,redhat7-POWERa,sles12-64a,sles15-64a,solaris114-64a,ubuntu1604-64a,ubuntu1604-32a,ubuntu1804-64a,windows10ent-64a,windows2012r2-64a,windows2016-64a,windows2019-64a

File: lib/src/facts/collection.cc [hunk 1]
```diff
@@ -163,46 +164,50 @@ namespace facter { namespace facts {
...
-                    }
-                    break;
-                }
+            try {
+                auto resolver = erf.get_resolver(path);
+                files.push_back(make_pair(path, resolver));
+            } catch (external::external_fact_no_resolver& e) {
+                LOG_WARNING("skipping file \"{1}\": {2}", path, e.what());
             }
             return true;
```
Comment (ciprianbadescu) [inline lib/src/facts/collection.cc]: as method is now void, check if this is still needed

Comment (ciprianbadescu) [reply lib/src/facts/collection.cc]: still needed, as this return is part of lambda function of `each_file`

Comment (GabrielNagy) [reply lib/src/facts/collection.cc]: I think you can drop the `true`

File: lib/tests/facts/windows/external_resolvers_factory.cc [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+            REQUIRE(erf.get_resolver(LIBFACTER_TESTS_DIRECTORY "/fixtures/facts/external/windows/execution/facts.bat"));
+        }
+    }
+    GIVEN("a PowerShell file extension") {
+        THEN("the file can be resolved") {
+            REQUIRE(erf.get_resolver(LIBFACTER_TESTS_DIRECTORY "/fixtures/facts/external/windows/powershell/facts.ps1"));
+        }
+    }
+    GIVEN("a non executable file without extension") {
+        THEN("should trow no resolver exception") {
```
Comment (GabrielNagy) [inline lib/tests/facts/windows/external_resolvers_factory.cc]: s/trow/throw/

File: lib/tests/facts/external_resolvers_factory.cc [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+#include <facter/facts/external_resolvers_factory.hpp>
+#include "../fixtures.hpp"
+
+using namespace std;
+using namespace facter::facts;
+
+SCENARIO("checking common external file resolvers factory") {
+    external_resolvers_factory erf;
+    GIVEN("a file with unknown extension") {
+        THEN("should trow no resolver exception") {
```
Comment (GabrielNagy) [inline lib/tests/facts/external_resolvers_factory.cc]: s/trow/throw/

**Pull request merged**